myia/ir.py

```diff
@@ -1,7 +1,142 @@
-"""Intermediate representation definition."""
+"""Intermediate representation definition.
+
+Myia's main intermediate representation (IR) is a graph-based version of ANF.
+Each function definition (lambda) is defined as a graph, consisting of a series
+of function applications.
+
+A function can be applied to a node from another funtion's graph; this
+implicitly creates a nested function. Functions are first-class objects, so
+returning a nested function creates a closure.
+
+"""
+from enum import Enum, auto
+from typing import List, Union, Dict
+
+LiteralType = Union[str, int, float, 'PrimitiveOperation', 'Graph', None]
+
+
+class Graph:
+    """A function graph.
+
+    Attributes:
+        output: The `Node` whose value will be returned by this function.
+
+    """
+
+    def __init__(self, output: 'Node') -> None:
+        """Construct a graph."""
+        self.output = output
+        self.debug: Dict = {}
 
 
 class Node:
-    """A node in the graph-based ANF IR."""
+    """A node in the graph-based ANF IR.
+
+    There are three types of nodes: Values, which are the result of function
+    applications; parameters; and constants such as numbers and functions.
+
+    Attributes:
+        inputs: If the node is a value, the first node input is the function to
+            apply, followed by the arguments.
+        value: The value of this node, if it is a constant.
+        graph: The function definition graph that this node belongs to for
+            values and parameters.
+        debug: A dictionary with debug information about this node e.g. a
+            human-readable name and the Python source code.
+
+    """
+
+    def __init__(self) -> None:
+        """Construct a node."""
+        self.inputs: List[Node] = []
+        self.value: LiteralType = None
```

Comment:
Should `inputs` and `value` be in the base class, if they are only used in subclasses? I suppose having `inputs` standard would make traversal a tad easier. As for `value`, if it's in the base class, I think `None` is a bad setting to signify that there is no value, because that could be a legitimate value (it also promotes the bad habit of writing `if node.value:`). I would prefer a singleton object like `UNKNOWN`.

Reply:
Yeah, the only reason I added `inputs` in the base class is so that general code using `for node in inputs` can be written. I agree that `None` is a bad default, but maybe the singleton should be `PARAMETER` for the `Parameter` class instead?

Reply:
Yeah, or some generic `NO_VALUE`. Note that in some cases you might want to create placeholder `Function` nodes for a graph while it is being built, so even `Constant` nodes may temporarily lack a valid value.